### PR TITLE
plutus-ir/plutus-tx: use force/delay

### PR DESCRIPTION
**Pull request by michaelpj**
Alternative to https://github.com/input-output-hk/plutus/pull/3640 using type abstractions.

I think let's do this for now, it's less invasive and the patch for the other one is still there if we want it.

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-tx-plugin/src/PlutusTx/Compiler/Laziness.hs [hunk 1]
```diff
@@ -42,7 +38,10 @@ delayVar (PIR.VarDecl () n ty) = do
 force
     :: CompilingDefault uni fun m
     => PIRTerm uni fun -> m (PIRTerm uni fun)
-force thunk = PIR.Apply () thunk <$> compileExpr (GHC.Var GHC.unitDataConId)
+force thunk = do
+    a <- liftQuote (freshTyName "dead")
```
Comment (kwxm) [inline plutus-tx-plugin/src/PlutusTx/Compiler/Laziness.hs:42]: It seems a bit morbid having all this death everywhere. I supposed it vanishes during erasure though, so at least it won't show up in the final output.

File: plutus-tx/src/PlutusTx/IsData/TH.hs [hunk 2]
```diff
@@ -90,7 +90,7 @@ unsafeReconstructCase (TH.ConstructorInfo{TH.constructorName=name, TH.constructo
             [|
                 let indexMatchCase = $(handleList argNames argsExpr)
                     fallthrough = $kont
-                in BI.ifThenElse ($ixExpr `BI.equalsInteger` (index :: Integer)) (const indexMatchCase) (const fallthrough) ()
+                in BI.ifThenElse ($ixExpr `BI.equalsInteger` (index :: Integer)) (const indexMatchCase) (const fallthrough) BI.unitval
```
Comment (kwxm) [inline plutus-tx/src/PlutusTx/IsData/TH.hs:93]: :+1:

File: plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs [hunk 3]
```diff
@@ -28,25 +28,50 @@ Since we are constructing a global substitution, so we need globally unique
 names to avoid clashes.
 -}
 
+{- Note [Using unit versus force/delay]
+We have force/delay in PIR, so we can use that for non-strict bindings, which is quite nice.
```
Comment (kwxm) [inline plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs:32]: I think this comment's outdated now.

**Review by kwxm — APPROVED** [review]
No objections, and it is simpler.  Devil's advocate: this works because of the (maybe debatable) choices we made about the semantics of type abstraction (Do we have the value restriction? Do we allow computation under Λ?), so in the unlikely event that we ever change that this might break.

**Comment by kwxm** [comment]
Grrr.  I ran the benchmarks but that was pointless because the scripts aren't regenerated!

**Comment by kwxm** [comment]
The scripts for this PR are in fact a little smaller than some I generated from the previous PR yesterday. That's presumably because of #3654.

**Comment by kwxm** [comment]
I regenerated the scripts for this PR and they're 1-2% slower than the ones generateddin the earlier PR, but still about 5% faster than some scripts I generated in master yesterday. Presumably that's because of other stuff that's happened in the meantime, since the erased scripts should be pretty much the same.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs [hunk 4]
```diff
@@ -28,25 +28,50 @@ Since we are constructing a global substitution, so we need globally unique
...
+
+However, we retain the *option* to use unit-lambdas instead, since we rely on this pass to
+handle recursive, non-function bindings and give them function types. `delayed x` is not a
+function type but `() -> x` is!
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs:36]: > Given this, can we have some optimization here as well?

In some cases. That would let us handle polymorphic functions, but not recursive values which are not functions! An edge case, true, but one that Haskell supports, and that this trick also handles. There's a test for it too.

> By the way, does this transformation change the semantics of a program?

No. PIR has non-strict let bindings, this turns them into strict let-bindings that behave the same.

I think what you're asking is "what do we do with Haskell let-bindings?". The answer is that we turn them into non-strict let-bindings unless we can tell the RHS is a value. So we do actually preserve some of the non-strictness of Haskell, rather than interpreting it strictly. I don't 100% remember why I decided to do this, but I think it was hard to avoid random things unexpectedly breaking if you didn't

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs:36]: Thanks, makes perfect sense.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs [hunk 5]
```diff
@@ -28,25 +28,50 @@ Since we are constructing a global substitution, so we need globally unique
...
+    -- See Note [Using unit versus force/delay]
+    let transformation = if useUnit then strictifyBindingWithUnit else strictifyBinding
+    in transformMOf termSubterms (traverseOf termBindings transformation)
 
 strictifyBinding
     :: (MonadState (Substs uni fun a) m, MonadQuote m)
     => Binding TyName Name uni fun a -> m (Binding TyName Name uni fun a)
 strictifyBinding = \case
+    TermBind x NonStrict (VarDecl x' name ty) rhs -> do
+        let ann = x
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs:62]: No reason.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs [hunk 6]
```diff
@@ -28,25 +28,50 @@ Since we are constructing a global substitution, so we need globally unique
...
+
+        a <- freshTyName "dead"
+        -- See Note [Compiling non-strict bindings]
+        modify $ Map.insert name $ TyInst ann (Var ann name) (TyForall ann a (Type ann) (TyVar ann a))
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/NonStrict.hs:66]: It would be nice, but I'd have to introduce the binding somewhere. This is simpler.
